### PR TITLE
sst: compress bloom filters

### PR DESCRIPTION
**Pull request by palash25**
Fixes: #152

### CODE REVIEW

**Comment by criccomini** [comment]
@palash25 This looks great. Only ask: Mind putting in a test to verify the filter's are getting compressed and decompressed properly?

**Comment by palash25** [comment]
thanks, i will update the tests today

**Comment by palash25** [comment]
hi @criccomini I made changes to the existing tests as filter compression and write path would already be covered by them I just added asserts to check if decompression was working fine it always passes this subtest `test_compression_inner`  but this one `test_compression_using_sst_info` always fails. i did some print debugging, found a few things but I am still stuck so needed a bit of help

first i checked if somehow the dummy compression was being used to decompress but that wasn't the case it was using the correct compression codec

i printed out the errors for each test failure they were the following for Snappy : `Err(Literal { len: 5, src_len: 3, dst_len: 5 })`   For Zlib: `Custom { kind: InvalidInput, error: "corrupt deflate stream" }`, For Lz4: `LiteralOutOfBounds` and  For Zstd: `Error decompressing Zstd data: Custom { kind: UnexpectedEof, error: "incomplete frame" }`

so I decided to check the compressed bytes and the sstableinfo in both the test cases and compare them if they differ but they didn't

Both  `test_compression_inner`   and `test_compression_using_sst_info` showed the following data when i printed it out 
```
SsTableInfoOwned { data: b"\x14\0\0\0\x10\0,\0\x08\0\x0c\0\x14\0\x1c\0$\0\x07\0\x10\0\0\0\0\0\0\x01$\0\0\0+\0\0\0\0\0\0\0.\0\0\0\0\0\0\0$\0\0\0\0\0\0\0\x05\0\0\0\0\0\0\0\x04\0\0\0key1" }

Block [b"&<\0\x04key1\0\0\0\x06value1\x05\x10\02\x15\x10\x182\0\0\0\x10\0\x02\xd0\x97\xff\x89\x05\x10\0\x06\x13\x84\x9b4\x04\x0c\0\x01\x01D\x06\0\x08\0\x04\0\x06\0\0\0\x04\0\0\0\x01\0\0\0\x01\x18\x08\x08\0\x08\x05\x12\t\x06$\0\0\x04\0\0\0key1\x14\0\0\0\x10\0,\0\x08\0\x0c\0\x14\0\x1c\0$\0\x07\0\x10\0\0\0\0\0\0\x01$\0\0\0+\0\0\0\0\0\0\0.\0\0\0\0\0\0\0$\0\0\0\0\0\0\0\x05\0\0\0\0\0\0\0\x04\0\0\0key1\x93\xad(\x93\0\0\0Y"]
```

The block data here is the is what we get at the end of the [`EncodedSsTableBuilder.build` method](https://github.com/slatedb/slatedb/blob/d005fc99a463a5a62a5293d05c4b5b4c88f9897d/src/sst.rs#L480)

If the compressed data is same in both cases I am not sure why the decompression fails in the second test, is there something that happens between the time when compressed data is written in store and then read back again?

the only two lines different between the test cases are these I don't see how they might affect decompression

```rust
let format = SsTableFormat::new(4096, 0, Some(dummy_codec));
let table_store = TableStore::new(object_store, format, root_path);
```

do you have any pointers on where to look next?

**Comment by palash25** [comment]
i wasn't able to work on this yesterday, i will take another look today

File: src/sst.rs [hunk 1]
```diff
@@ -431,7 +446,11 @@ impl<'a> EncodedSsTableBuilder<'a> {
             let filter = Arc::new(self.filter_builder.build());
             let encoded_filter = filter.encode();
             filter_len = encoded_filter.len();
```
Comment (criccomini) [inline src/sst.rs:448]: I believe this is the issue. The filter length is set to the encoded bytes _before_ they're compressed. The length should be set _after_ compression. I moved this line above the linen 453 (`buf.put(...)`) and the tests passed.

**Comment by criccomini** [comment]
I committed the re-order and merged. Thanks! 😄

File: src/sst.rs (hunk 1 above)
Comment (palash25) [reply src/sst.rs]: ah thanks

**Pull request merged**